File: src/main/java/kg/musabaev/cluserizator/TextFieldTreeCellImpl.java
package kg.musabaev.cluserizator;

import javafx.scene.control.TextField;
import javafx.scene.control.TreeCell;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyEvent;

public class TextFieldTreeCellImpl extends TreeCell<SeoKeywordModel> {

    private TreeCellGraphic graphicImpl;

    public TextFieldTreeCellImpl() {
    }

    @Override
    public void startEdit() {
        super.startEdit();

        if (graphicImpl == null) {
            createTextField();
        }
        setText(null);
        setGraphic(graphicImpl);
        graphicImpl.getInput().requestFocus();
//        textField.selectPositionCaret(textField.getText().length());
    }

    @Override
    public void cancelEdit() {
        super.cancelEdit();
        setText(getItem().getKeyword());
        setGraphic(getTreeItem().getGraphic());
        this.requestFocus();
    }

    @Override
    public void updateItem(SeoKeywordModel item, boolean empty) {
        super.updateItem(item, empty);

        if (empty) {
            setText(null);
            setGraphic(null);
        } else {
            if (isEditing()) {
                if (graphicImpl != null) {
                    graphicImpl.getInput().setText(getString());
                }
                setText(null);
                setGraphic(graphicImpl);
            } else {
                setGraphic(getTreeItem().getGraphic());
            }
        }
    }

    private void createTextField() {
        graphicImpl = new TreeCellGraphic(getItem().getId(), getItem().getKeyword(), getItem().getOtherMetas()[1]);
        graphicImpl.getInput().setOnKeyReleased((KeyEvent t) -> {
            if (t.getCode() == KeyCode.ENTER) {
                getItem().setKeyword(graphicImpl.getInput().getText());
                commitEdit(getItem());
            } else if (t.getCode() == KeyCode.ESCAPE) {
                cancelEdit();
            }
        });
    }

    private String getString() {
        return getItem() == null ? "<<null>>" : getItem().getKeyword();
    }
}

File: src/main/kotlin/kg/musabaev/cluserizator/TextFieldTreeCellImpl.kt
package kg.musabaev.cluserizator

import javafx.event.EventHandler
import javafx.scene.control.TreeCell
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyEvent

class TextFieldTreeCellImpl : TreeCell<SeoKeywordModel?>() {
    private var graphicImpl: TreeCellGraphic? = null

    override fun startEdit() {
        super.startEdit()

        if (graphicImpl == null) {
            createTextField()
        }
        text = null
        graphic = graphicImpl
        graphicImpl?.input?.requestFocus()
        //        textField.selectPositionCaret(textField.getText().length());
    }

    override fun cancelEdit() {
        super.cancelEdit()
        text = item!!.getKeyword()
        graphic = treeItem.graphic
        this.requestFocus()
    }

    override fun updateItem(item: SeoKeywordModel?, empty: Boolean) {
        super.updateItem(item, empty)

        if (empty) {
            text = null
            graphic = null
        } else {
            if (isEditing) {
                if (graphicImpl != null) {
                    graphicImpl!!.input.text = string
                }
                text = null
                graphic = graphicImpl
            } else {
                graphic = treeItem.graphic
            }
        }
    }

    private fun createTextField() {
        graphicImpl = TreeCellGraphic(item!!.getId(), item!!.getKeyword(), item!!.otherMetas[1])
        graphicImpl!!.input.onKeyReleased = EventHandler { t: KeyEvent ->
            if (t.code == KeyCode.ENTER) {
                item!!.setKeyword(graphicImpl!!.input.text)
                commitEdit(item)
            } else if (t.code == KeyCode.ESCAPE) {
                cancelEdit()
            }
        }
    }

    private val string: String
        get() = if (item == null) "<<null>>" else item!!.getKeyword()
}